OpenRA.Mods.AS/Traits/BotModules/PowerDownBotASModule.cs

```diff
@@ -0,0 +1,155 @@
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Manages AI powerdown.")]
+	public class PowerDownBotASModuleInfo : ConditionalTraitInfo
+	{
+		[Desc("Delay (in ticks) between toggling powerdown")]
+		public readonly int Interval = 150;
+
+		[Desc("ConditionToken that powerdown")]
+		public readonly string PowerdownCondition = "powerdown";
+
+		public override object Create(ActorInitializer init) { return new PowerDownBotASModule(init.Self, this); }
+	}
+
+	public class PowerDownBotASModule : ConditionalTrait<PowerDownBotASModuleInfo>, IBotTick
+	{
+		readonly World world;
+		readonly Player player;
+		PowerManager playerPower;
+		int toggleTick;
+		readonly Func<Actor, bool> isToggledBuildingsValid;
+		List<BuildingPowerWrapper> toggledBuildings;
+
+		class BuildingPowerWrapper
+		{
+			public int PowerChanging;
+			public Actor Actor;
+
+			public BuildingPowerWrapper(Actor a, int p)
+			{
+				Actor = a;
+				PowerChanging = p;
+			}
+		}
+
+		public PowerDownBotASModule(Actor self, PowerDownBotASModuleInfo info)
+			: base(info)
+		{
+			world = self.World;
+			player = self.Owner;
+			toggledBuildings = new List<BuildingPowerWrapper>();
+			isToggledBuildingsValid = a => a.Owner == self.Owner && !a.IsDead && a.IsInWorld && GetTogglePowerChanging(a) < 0;
+		}
+
+		protected override void Created(Actor self)
+		{
+			// Special case handling is required for the Player actor.
+			// Created is called before Player.PlayerActor is assigned,
+			// so we must query player traits from self, which refers
+			// for bot modules always to the Player actor.
+			playerPower = self.TraitOrDefault<PowerManager>();
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			toggleTick = world.LocalRandom.Next(0, Info.Interval);
+			toggledBuildings = new List<BuildingPowerWrapper>();
+		}
+
+		int GetTogglePowerChanging(Actor a)
+		{
+			var powerChangingIfToggled = 0;
+			var powerTarit = a.TraitsImplementing<Power>().Where(t => !t.IsTraitDisabled).ToArray();
+			var powerMulTrait = a.TraitsImplementing<PowerMultiplier>().ToArray();
+			if (powerTarit.Any())
+			{
+				powerChangingIfToggled = powerTarit.Sum(p => p.Info.Amount) * (powerMulTrait.Sum(p => p.Info.Modifier) - 100) / 100;
+				if (powerMulTrait.Where(t => !t.IsTraitDisabled).Any())
+					powerChangingIfToggled = -powerChangingIfToggled;
+			}
+
+			return powerChangingIfToggled;
+		}
+
+		IEnumerable<Actor> GetToggleableBuildings(IBot bot)
+		{
+			var toggleable = bot.Player.World.ActorsHavingTrait<ToggleConditionOnOrder>(t => !t.IsTraitDisabled && !t.IsTraitPaused && t.Info.Condition == Info.PowerdownCondition)
```

Comment:
This sounds too expensive to me and I already see issues with BotTicks being expensive in my own tests. This also gives a lesser amount of control for modders to use this feature. I'd really vote for creating a dummy trait (say `EligibleForBotPowerDown` which requires `ToggleConditionOnOrder` and then this entire line can just be simplified to `bot.Player.World.ActorsHavingTrait<EligibleForBotPowerDown>()`.

Reply:
This trait might not even be a dummy one, actually because it can then implement the `INotifyBeingCaptured` interface to force an unpowering immediately if it was captured by a bot while powered down.

Reply:
I guess this one can be better: add all building/actors can be powerdown when they are created. But it will be a big project.

Reply:
Those would then need updates when they change ownership as well. Having the dummy trait is a good middle ground IMO.

Reply:
And I also need to ask here, I think `t => !t.IsTraitDisabled && !t.IsTraitPaused && t.Info.Condition == Info.PowerdownCondition` is not that expensive and many bot module use it casually(SquadManagerBotModule,  UnitBuilderBotModuleInfo, etc), is there something missing?

Reply:
`t.Info.Condition == Info.PowerdownCondition` is a string comparison which is fairly expensive (Plus the dummy trait simplifies things a lot with the implicit assumptions that it is appropriate so you don't have to check for the followup traits like IsBuilding as well). You're also doing this lookup in every tick compared to most other bot modules which have an interval check to only do these updates periodically. Adding this latter is actually a fairly good improvement as well.

Reply:
to fix this, I have another idea which is just to simply remove it, but upon doing this will request modders/gamers make "toggle" trait only used for powerdown, the bot trait here will take all actor with toggle trait into consideration for power control and I don't know if someone here will use this toggle trait to do something more than power control stuff (although they can actually use deploy for that)

Reply:
Try my idea.
Maybe a dummy trait here is a good point for extension but probably we should suggest those who mod on OpenRA only use toggle for power control.

Reply:
Okay, using `ToggleConditionOnOrder` as the tag trait is fine by me.